wgengine/magicsock/magicsock_linux.go

```diff
@@ -0,0 +1,233 @@
+// Copyright (c) 2022 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package magicsock
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/netip"
+	"unsafe"
+
+	"golang.org/x/net/bpf"
+	"golang.org/x/sys/unix"
+	"tailscale.com/disco"
+	"tailscale.com/envknob"
+	"tailscale.com/types/ipproto"
+	"tailscale.com/types/key"
+	"tailscale.com/util/clientmetric"
+)
+
+const (
+	udpHeaderSize          = 8
+	ipv6HeaderSize         = 40
+	ipv6FragmentHeaderSize = 8
+)
+
+// Debug tweakables
+var (
+	// Enable/disable using AF_PACKET sockets
+	debugDisableAF_PACKET = envknob.Bool("TS_DEBUG_DISABLE_AF_PACKET")
+)
+
+// Convert our disco magic number into a uint32 and uint16 to test
+// against. We panic on an incorrect length here rather than try to be
+// generic with our BPF instructions below.
+//
+// Note that BPF uses network byte order (big-endian) when loading data
+// from a packet, so that is what we use to generate our magic numbers.
+var magic1, magic2 = (func() (uint32, uint16) {
```

Comment:
two consts or two vars without the func literal at least? no need to check the disco length

Reply:
Replaced with a const magic and a test to enforce that it matches the disco magic in the disco packet.